Fix issues #7 and #24 for priority inversion in firing timed transitions

airabinovich opened the pull request:
New class PriorityBinaryLock, a lock with two priority levels.
MonitorManager and FairQueue now use PriorityLock instead of Semaphore for their queues.
VarCondQueue adds a new method for high priority sleep.
Added tests for PriorityLock and FairQueue.

Tests for actual priority inversion are needed.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/airabinovich/java_petri_engine/27)

<!-- Reviewable:end -->

airabinovich committed:
Fix issues #7 and #24 for priority inversion in firing timed transitions.
New class PriorityBinaryLock, a lock with two priority levels.
MonitorManager and FairQueue now use PriorityLock instead of Semaphore for their queues.
VarCondQueue adds a new methor for high priority sleep.
Added tests for PriorityLock and FairQueue.

mauriludemann commented:
*[src/monitor_petri/FairQueue.java, line 15 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjUT9J3BO-AImgwUZc:-KUjUT9J3BO-AImgwUZd:bwwipqm) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L15)):*
> ```Java
> 	public FairQueue(){
> 		lock = new PriorityBinaryLock();
> 		// take the lock so no thread calling sleep can actually take it and continue
> ```

Comment unclear

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
*[src/monitor_petri/MonitorManager.java, line 315 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjVZG7CM-lvepmym7m:-KUjVZG7CM-lvepmym7n:b-cintxw) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/MonitorManager.java#L315)):*
> ```Java
> 				while(isTimed && !insideTimeSpan){
> 					if(transitionSpan.isBeforeTimeSpan(fireAttemptTime) && !transitionSpan.anySleeping()){
> 						// The calling thread came before time span, and there is nobody sleeping in the transition,
> ```

"sleeping in the transition"
In the transition or by itself?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
*[src/monitor_petri/PriorityBinaryLock.java, line 10 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjXQrGjmUU467eUEXI:-KUjXQrGjmUU467eUEXJ:bxacbuh) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/PriorityBinaryLock.java#L10)):*
> ```Java
> /**
>  * A lock with two priority levels.
>  * There is only one permit available, thus the binary.
> ```

"thus the binary"
what does it mean?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/monitor_petri/FairQueue.java, line 15 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjUT9J3BO-AImgwUZc:-KUlylV8ffejIWonq-VZ:b-xcv3zi) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L15)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> Comment unclear
</details>
I couldn't figure out another way to put it.
Since "lock" is a binary lock that can't be initialized as taken, the main thread takes it and never releases. That'll make any thread calling sleep() (and so trying to take the lock) to sleep and not actually take the lock and simply continue.
And because there is no restriction on who takes and releases the lock, a thread calling wakeUp() will release the lock and the first thread (in order of priority and time) waiting will take it if there's any available

---

*[src/monitor_petri/MonitorManager.java, line 315 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjVZG7CM-lvepmym7m:-KUlyM3jfziX5sNEKmsr:b-4wtny9) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/MonitorManager.java#L315)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> "sleeping in the transition"
> In the transition or by itself?
</details>
you got me, "sleeping waiting for this transition" is ok?

---

*[src/monitor_petri/PriorityBinaryLock.java, line 10 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjXQrGjmUU467eUEXI:-KUlya7FK2ivsR2MK7C-:b-cctfvj) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/PriorityBinaryLock.java#L10)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> "thus the binary"
> what does it mean?
</details>
maybe it can be more clear as "thus, the 'binary' in the class' name"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

juanjoarce7456 commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/monitor_petri/FairQueue.java, line 15 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjUT9J3BO-AImgwUZc:-KUmEZ1tltFoDZb7BIQ3:b-7bqt4g) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L15)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> I couldn't figure out another way to put it.
> Since "lock" is a binary lock that can't be initialized as taken, the main thread takes it and never releases. That'll make any thread calling sleep() (and so trying to take the lock) to sleep and not actually take the lock and simply continue.
> And because there is no restriction on who takes and releases the lock, a thread calling wakeUp() will release the lock and the first thread (in order of priority and time) waiting will take it if there's any available
</details>
maybe it could be something more general like "Setting lock permits to 0"

---

*[src/monitor_petri/FairQueue.java, line 43 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmFpgGZObmVhlxqjhM:-KUmFpgGZObmVhlxqjhN:b-o9k25j) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L43)):*
> ```Java
> 	}
> 	
> 	public int getHighPriorityThreadsSLeeping(){
> ```

typo in Sleeping. "SL"
I think it should be getHighPriorityThreadsSleepingCount

---

*[src/monitor_petri/FairQueue.java, line 47 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmG26mcW1v1sNNs1ER:-KUmG26mcW1v1sNNs1ES:bjquptk) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L47)):*
> ```Java
> 	}
> 	
> 	public int getLowPriorityThreadsSLeeping(){
> ```

same as before

---

*[src/monitor_petri/MonitorManager.java, line 114 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmGEY5IAAecV4eOEbv:-KUmGEY5IAAecV4eOEbw:b7du2kd) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/MonitorManager.java#L114)):*
> ```Java
> 			throw new IllegalTransitionFiringError("An automatic transition has tried to be fired manually");
> 		}
> 		boolean releaseLock = true;
> ```

Are the permits to release never  > 1?
I think they are not, just asking.

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 41 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmRcvGJHX8Dd_nuQ3X:-KUmRcvGJHX8Dd_nuQ3Y:b-nr3p3f) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/test/unit_tests/PriorityBinaryLockTestSuite.java#L41)):*
> "is gets" 
shouldn't it be be "gets"?
same on every test

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 320 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmQwxGropT386WYqJ2:-KUmQwxGropT386WYqJ3:b3b79d9) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/test/unit_tests/PriorityBinaryLockTestSuite.java#L320)):*
> ```Java
> 	 * <li> And l0's high priority queue is empty </li>
> 	 * <li> When 50 threads try to take l0 with low priority </li>
> 	 * <li> And 50 threads try to take l0 with low priority </li>
> ```

high priority

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/monitor_petri/FairQueue.java, line 15 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjUT9J3BO-AImgwUZc:-KUmVCSa5xD5Ih5H4o9V:b-uzzigq) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L15)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> maybe it could be something more general like "Setting lock permits to 0"
</details>
It's a binary lock, technically it doesn't have "pemits"

---

*[src/monitor_petri/FairQueue.java, line 43 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmFpgGZObmVhlxqjhM:-KUmVMCUgLjR4BF8N6Kg:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L43)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> typo in Sleeping. "SL"
> I think it should be getHighPriorityThreadsSleepingCount
</details>
Done.

---

*[src/monitor_petri/FairQueue.java, line 47 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmG26mcW1v1sNNs1ER:-KUmVMwcp2SgWOnSVw5E:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L47)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> same as before
</details>
Done.

---

*[src/monitor_petri/MonitorManager.java, line 114 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmGEY5IAAecV4eOEbv:-KUmUvXq-kdua1CJpokv:bh1jtjk) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/MonitorManager.java#L114)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> Are the permits to release never  > 1?
> I think they are not, just asking.
</details>
No, never. If it was that way, two threads might enter the monitor simultaneously

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 41 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmRcvGJHX8Dd_nuQ3X:-KUmV5ldpbD-fIvZOGtR:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/test/unit_tests/PriorityBinaryLockTestSuite.java#L41)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> > "is gets" 
> shouldn't it be be "gets"?
> same on every test
</details>
Done.

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 320 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUmQwxGropT386WYqJ2:-KUmV85LxswhwkcYoIX_:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/test/unit_tests/PriorityBinaryLockTestSuite.java#L320)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> high priority
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/monitor_petri/MonitorManager.java, line 315 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjVZG7CM-lvepmym7m:-KUtq-GuLD-6UQeqkuGb:b-hfghm4) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/MonitorManager.java#L315)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> you got me, "sleeping waiting for this transition" is ok?
</details>
Sounds good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich committed:
Merge branch 'dev' into fix_priorityInversionWhenThreadWakesUp

airabinovich committed:
Added tests for priority inversion when firing timed transitions and multiple threads try to fire the same transitions.
Fixed MonitorManager for both cases.
Minor modifications to PriorityBinaryLock.

airabinovich commented:
Review status: 2 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/monitor_petri/MonitorManager.java, line 315 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjVZG7CM-lvepmym7m:-KUz3bh-f1d2u2Y2WylF:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/MonitorManager.java#L315)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> Sounds good
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
Review status: 2 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/monitor_petri/FairQueue.java, line 15 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjUT9J3BO-AImgwUZc:-KV2zlNqq-4tBxI_3dz7:b6e9o9u) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L15)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> It's a binary lock, technically it doesn't have "pemits"
</details>
did you fix it?

---

*[src/monitor_petri/PriorityBinaryLock.java, line 54 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV30XFTaA049dPnWM-g:-KV30XFTaA049dPnWM-h:bczd4fc) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/src/monitor_petri/PriorityBinaryLock.java#L54)):*
> ```Java
> 	/**
> 	 * Tries to lock this PriorityLock with low priority.
> 	 * Takes the lock if available and returns immediately,
> ```

what thing returns inmediately?

---

*[src/monitor_petri/PriorityBinaryLock.java, line 65 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV377NVzPkj8S6m_asI:-KV377NWMbQqk9ZhrtCv:bnj7hl0) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/src/monitor_petri/PriorityBinaryLock.java#L65)):*
> ```Java
> 	/**
> 	 * Tries to lock this PriorityLock with the specified priority.
> 	 * Takes the lock if available and returns immediately,
> ```

same as above

---

*[test/unit_tests/FairQueueTestSuite.java, line 91 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3A3lCzQa1hARG_RQB:-KV3A3lCzQa1hARG_RQC:b-r9w66c) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/FairQueueTestSuite.java#L91)):*
> Given 3 threads go to sleep
with low priority

---

*[test/unit_tests/FairQueueTestSuite.java, line 105 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3AEmrW-fVl5UkttlK:-KV3AEmrW-fVl5UkttlL:b2isd6m) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/FairQueueTestSuite.java#L105)):*
> ```Java
> 		}
> 		
> 		for(int i = 3; i < 5; i++){
> ```

3 to 5 ^^

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 561 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3AvmzFMbZYMbg0i7F:-KV3AvmzFMbZYMbg0i7G:bbmg2tl) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L561)):*
> ```Java
> 	* <li> When thread th0 tries to fire t0 at time tj, ti < tj < (ti+a) </li>
> 	* <li> And th0 sleeps on its own waiting for t0's timespan </li>
> 	* <li> And 500 threads try to fire t2 before th1 wakes up</li>
> ```

500, good number !

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV3BNDmxSuDG4VVcL_6:byz8fg7) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
> ```Java
> 		
> 		th0.start();
> 		
> ```

maybe you could add an assert, testing that th0 is waiting for the transition?

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV3CIH4BT-wOVV2laVL:bj8blcs) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
> ```Java
> 		
> 		th0.start();
> 		
> ```

same as above 

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 207 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3E4ZhXLv0_Sq5v1Hy:-KV3E4ZhXLv0_Sq5v1Hz:b-d1mvqv) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L207)):*
> ```Java
> 
> 		try{
> 			ft1.get(1, TimeUnit.MILLISECONDS);
> ```

is this behavior expected? when th0 finishes its task, makes an unlock or not?

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 272 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3FGgKNXATy1GnJEsc:-KV3FGgKNXATy1GnJEsd:be4q8v7) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L272)):*
> ```Java
> 	 * <li> And l0's low priority queue is empty </li>
> 	 * <li> And l0's high priority queue is empty </li>
> 	 * <li> When thread th0 tries to lock l0 with low priority </li>
> ```

Do you mean "high" priority?

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 338 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3Fv5pxHJD9w3p108B:-KV3Fv5pxHJD9w3p108C:bn9h61y) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L338)):*
> ```Java
> 		
> 		boolean isHighPriority = true;
> 		for(int i = 0; i < 100; i++){
> ```

you didn't want to do two "for"? ^^

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 367 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3GdDuPReG-cDdjc6O:-KV3GdDuPReG-cDdjc6P:b-437imy) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L367)):*
> ```Java
> 			try {
> 				// give the waken thread a little time to actually get out of the lock
> 				Thread.sleep(1);
> ```

Too much time :O

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich commented:
Review status: 2 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/monitor_petri/FairQueue.java, line 15 at r1](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KUjUT9J3BO-AImgwUZc:-KV58ajRlmcmV8cvOaU_:bkool2j) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/3483ee68520cd7409c4bdde6db5776dff448f476/src/monitor_petri/FairQueue.java#L15)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> did you fix it?
</details>
read the new comment

---

*[src/monitor_petri/PriorityBinaryLock.java, line 54 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV30XFTaA049dPnWM-g:-KV58gXEs3QlOvJtsf-E:b-7dazcp) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/src/monitor_petri/PriorityBinaryLock.java#L54)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> what thing returns inmediately?
</details>
the program flow, the calling thread

---

*[src/monitor_petri/PriorityBinaryLock.java, line 65 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV377NVzPkj8S6m_asI:-KV5B5JgyjlF4oBUSi2m:bm97h5n) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/src/monitor_petri/PriorityBinaryLock.java#L65)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> same as above
</details>
read previous comment

---

*[test/unit_tests/FairQueueTestSuite.java, line 91 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3A3lCzQa1hARG_RQB:-KV59CH_rHIhdJrD_mNY:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/FairQueueTestSuite.java#L91)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> > Given 3 threads go to sleep
> with low priority
</details>
Done.

---

*[test/unit_tests/FairQueueTestSuite.java, line 105 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3AEmrW-fVl5UkttlK:-KV599X0gXwkbAn_RAG9:bon15u1) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/FairQueueTestSuite.java#L105)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> 3 to 5 ^^
</details>
I don't see a problem here

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 561 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3AvmzFMbZYMbg0i7F:-KV59IaDLBGQD4wQZGRo:b-i5ablq) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L561)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> 500, good number !
</details>
I needed to make the race condition certain and happening every time

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV59TJe_Ek9WnpaHf2L:b-fr5y1h) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> maybe you could add an assert, testing that th0 is waiting for the transition?
</details>
there's no way to test that

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV5B7GJ0sqXliWf4TCx:bm97h5n) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> same as above 
</details>
read previous comment

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 207 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3E4ZhXLv0_Sq5v1Hy:-KV59w-psrzg-6ku_vdI:b-f2gmi0) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L207)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> is this behavior expected? when th0 finishes its task, makes an unlock or not?
</details>
it's not in ft0 so no, remember this is just the lock

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 272 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3FGgKNXATy1GnJEsc:-KV5AS2Cm0LD7gsZoJ2s:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L272)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> Do you mean "high" priority?
</details>
Done.

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 338 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3Fv5pxHJD9w3p108B:-KV5AXKDuPxA3gCDctDY:bfb63fs) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L338)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> you didn't want to do two "for"? ^^
</details>
this is for the threads to be generated in a mixed way and not all low priority together adn all high priority together

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 367 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3GdDuPReG-cDdjc6O:-KV5AlQWIzxUfx77vCxV:b-ad3uor) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L367)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> Too much time :O
</details>
just enough. without this there might be an issue

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich committed:
a few fixes for review comments

mauriludemann commented:
Review status: 2 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[test/unit_tests/FairQueueTestSuite.java, line 105 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3AEmrW-fVl5UkttlK:-KV5JfdyNQX6ZQkB4csL:b-2jhbbq) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/FairQueueTestSuite.java#L105)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> I don't see a problem here
</details>
It never was a problem :)

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV5JxdfeUCIBn1q_gj0:b-h9z3p1) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> there's no way to test that
</details>
t0.getTimeSpan().anySleeping()

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV5KErCSMTfBbUhaJZ5:b-h9z3p1) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> read previous comment
</details>
t0.getTimeSpan().anySleeping()

---

*[test/unit_tests/PriorityBinaryLockTestSuite.java, line 207 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3E4ZhXLv0_Sq5v1Hy:-KV5KNG-d6UjAkIibiFW:b-3943ik) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/PriorityBinaryLockTestSuite.java#L207)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> it's not in ft0 so no, remember this is just the lock
</details>
Ok, is the lock and no a firing, good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich commented:
Review status: 2 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV5LcojxFFLcm69zRHe:b-78rjnu) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> t0.getTimeSpan().anySleeping()
</details>
anySleeping() doesn't tell if it's th0 or another thread

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV5LlVea5hTF2lmY_mu:bsg9y7n) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> t0.getTimeSpan().anySleeping()
</details>
same as above, I can't be sure it's th0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
Review status: 2 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV5MQ31cvhkgUyEmqM2:b9ra27d) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> anySleeping() doesn't tell if it's th0 or another thread
</details>
Yes, but at this time you are firing only t0 with th0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
Review status: 2 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV5M_pONrU6QsTtq1dJ:b-ya81ng) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
<details><summary><i>Previously, airabinovich (Ariel Iván Rabinovich) wrote…</i></summary>
> same as above, I can't be sure it's th0
</details>
Same as above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

juanjoarce7456 commented:
Review status: 2 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV5m87ytF6jWnpAbOtr:b-ir2txa) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> Yes, but at this time you are firing only t0 with th0
</details>
I think mauri is right 

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 656 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV5nNJZjnhYH7sQDZ57:-KV5nNJ_kdSNI5FJPJTD:b-5uzcn2) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L656)):*
> ```Java
> 	* <li> And th0 wakes up and fires t0 </li>
> 	* <li> And obs got only one message with t0's ID </li>
> 	* <li> And th0's final state is TERMIANTED </li>
> ```

typo: TERMIANTED 

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV5nbHXHgOdtjBnV6cn:b2v4cer) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
<details><summary><i>Previously, mauriludemann (Mauri) wrote…</i></summary>
> Same as above
</details>
Same as above (I just wanna be popular)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

juanjoarce7456 commented:
Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

airabinovich commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 603 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3BNDmxSuDG4VVcL_5:-KV5wZHn1DXzPKN9buA3:b-yhnmnh) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L603)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> I think mauri is right 
</details>
Ok, I'll add it

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 656 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV5nNJZjnhYH7sQDZ57:-KV5xllNZL5QhPT5SmA8:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L656)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> typo: TERMIANTED 
</details>
Done.

---

*[test/unit_tests/MonitorManagerTimeTestSuite.java, line 698 at r2](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-KV3CIH3Gbjwn5tqTgNM:-KV5y2bwk99kKGHsc4Dl:b-896fix) ([raw file](https://github.com/airabinovich/java_petri_engine/blob/deb2e78a433e7d3254e4bc2a67fe01fbdf9a4e28/test/unit_tests/MonitorManagerTimeTestSuite.java#L698)):*
<details><summary><i>Previously, juanjoarce7456 wrote…</i></summary>
> Same as above (I just wanna be popular)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 8 of 11 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27#-:-KV6316UarFbYLloReTs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mauriludemann commented:
Reviewed 2 of 7 files at r1, 6 of 9 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/airabinovich/java_petri_engine/27)*
<!-- Sent from Reviewable.io -->
